File: plot/src/main/java/net/fwitz/math/plot/in3d/Vertex.java
package net.fwitz.math.plot.in3d;

import static java.util.Objects.requireNonNull;

public class Vertex {
    private final Vector3 v;
    private final int rgb;

    public Vertex(Vector3 v, int rgb) {
        this.v = requireNonNull(v, "v");
        this.rgb = rgb;
    }

    public Vertex(double x, double y, double z, int rgb) {
        this(new Vector3(x, y, z), rgb);
    }

    public Vector3 v() {
        return v;
    }

    public double x() {
        return v.x();
    }

    public double y() {
        return v.y();
    }

    public double z() {
        return v.z();
    }

    public int rgb() {
        return rgb;
    }

    @Override
    public boolean equals(Object o) {
        return this == o || (o instanceof Vertex && equalTo((Vertex) o));
    }

    private boolean equalTo(Vertex other) {
        return rgb == other.rgb && v.equals(other.v);
    }

    @Override
    public int hashCode() {
        return v.hashCode() * 31 + rgb;
    }

    @Override
    public String toString() {
        return "Vertex[v=" + v + ",rgb=" + hexColor(rgb) + ']';
    }

    private static String hexColor(int rgb) {
        return String.format("#%02X%02X%02X",
                (rgb >> 16) & 0xFF,
                (rgb >> 8) & 0xFF,
                rgb & 0xFF);
    }
}

File: plot/src/main/kotlin/net/fwitz/math/plot/in3d/Vertex.kt
package net.fwitz.math.plot.in3d

import java.util.Objects

class Vertex(v: Vector3, rgb: Int) {
    private val v: Vector3
    private val rgb: Int

    constructor(x: Double, y: Double, z: Double, rgb: Int) : this(Vector3(x, y, z), rgb) {}

    fun v(): Vector3 {
        return v
    }

    fun x(): Double {
        return v.x()
    }

    fun y(): Double {
        return v.y()
    }

    fun z(): Double {
        return v.z()
    }

    fun rgb(): Int {
        return rgb
    }

    override fun equals(o: Any?): Boolean {
        return this === o || o is Vertex && equalTo(o)
    }

    private fun equalTo(other: Vertex): Boolean {
        return rgb == other.rgb && v == other.v
    }

    override fun hashCode(): Int {
        return v.hashCode() * 31 + rgb
    }

    override fun toString(): String {
        return "Vertex[v=" + v + ",rgb=" + hexColor(rgb) + ']'
    }

    companion object {
        private fun hexColor(rgb: Int): String {
            return String.format(
                "#%02X%02X%02X",
                rgb shr 16 and 0xFF,
                rgb shr 8 and 0xFF,
                rgb and 0xFF
            )
        }
    }

    init {
        this.v = Objects.requireNonNull(v, "v")
        this.rgb = rgb
    }
}